### PR TITLE
Fix kubelet cadvisor stats runtime panic

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
Fixing a kubelet runtime panic when the runtime returns incomplete data:

```
E0729 08:17:47.260393    5218 panic.go:115] "Observed a panic" panic="runtime error: index out of range [0] with length 0" panicGoValue="runtime.boundsError{x:0, y:0, signed:true, code:0x0}" stacktrace=<
        goroutine 174 [running]:
        k8s.io/apimachinery/pkg/util/runtime.logPanic({0x33631e8, 0x4ddf5c0}, {0x2c9bfe0, 0xc000a563f0})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0xbc
        k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x33631e8, 0x4ddf5c0}, {0x2c9bfe0, 0xc000a563f0}, {0x4ddf5c0, 0x0, 0x10000000043c9e5?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:82 +0x5e
        k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000ae08c0?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
        panic({0x2c9bfe0?, 0xc000a563f0?})
                runtime/panic.go:785 +0x132
        k8s.io/kubernetes/pkg/kubelet/stats.(*cadvisorStatsProvider).ImageFsStats(0xc000535d10, {0x3363348, 0xc000afa330})
                k8s.io/kubernetes/pkg/kubelet/stats/cadvisor_stats_provider.go:277 +0xaba
        k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).GarbageCollect(0xc000a3c820, {0x33631e8?, 0x4ddf5c0?}, {0x0?, 0x0?, 0x4dbca20?})
                k8s.io/kubernetes/pkg/kubelet/images/image_gc_manager.go:354 +0x1d3
        k8s.io/kubernetes/pkg/kubelet.(*Kubelet).StartGarbageCollection.func2()
                k8s.io/kubernetes/pkg/kubelet/kubelet.go:1472 +0x58
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000add110, {0x3330380, 0xc000afa300}, 0x1, 0xc0000ac150)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
        k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000add110, 0x45d964b800, 0x0, 0x1, 0xc0000ac150)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
        k8s.io/apimachinery/pkg/util/wait.Until(...)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:161
        created by k8s.io/kubernetes/pkg/kubelet.(*Kubelet).StartGarbageCollection in goroutine 1
                k8s.io/kubernetes/pkg/kubelet/kubelet.go:1470 +0x247
```

This commit fixes panics if:

- `len(imageStats.ImageFilesystems) == 0`
- `len(imageStats.ContainerFilesystems) == 0`
- `imageStats.ImageFilesystems[0].FsId == nil`
- `imageStats.ContainerFilesystems[0].FsId == nil`
- `imageStats.ImageFilesystems[0].UsedBytes == nil`
- `imageStats.ContainerFilesystems[0].UsedBytes == nil`

It also fixes the wrapped `nil` error for the check: `err != nil || imageStats == nil` in case that `imageStats == nil`.
#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
PTAL @kubernetes/sig-node-pr-reviews 

This seems to be critical for v1.31 @kubernetes/release-team-leads.

**Comment by saschagrunert** [comment]
cc @kannon92 @haircommander @bitoku @kwilczynski

**Comment by aojea** [comment]
can we add an unit test that at least return an incomplete imageStats ?

**Comment by saschagrunert** [comment]
> can we add an unit test that at least return an incomplete imageStats ?

Yep, done :+1:

File: pkg/kubelet/stats/cadvisor_stats_provider.go
Comment (kwilczynski) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:277]: To make it more explicit, perhaps.

```suggestion
		return nil, nil, fmt.Errorf("missing image stats: %+v", imageStats)
```

File: pkg/kubelet/stats/cadvisor_stats_provider_test.go
Comment (kwilczynski) [inline pkg/kubelet/stats/cadvisor_stats_provider_test.go:620]: Some linter does not seem to like this assertion.

**Comment by kwilczynski** [comment]
/hold

For the linter complaints.

**Comment by kwilczynski** [comment]
Thank you @saschagrunert for a switch fix! Much appreciated! :bow:

File: pkg/kubelet/stats/cadvisor_stats_provider_test.go
Comment (saschagrunert) [reply pkg/kubelet/stats/cadvisor_stats_provider_test.go]: Yes, fixed.

File: pkg/kubelet/stats/cadvisor_stats_provider.go
Comment (saschagrunert) [reply pkg/kubelet/stats/cadvisor_stats_provider.go]: Fixed as suggested.

**Comment by kwilczynski** [comment]
Related:

- https://github.com/kubernetes/kubernetes/pull/126205
- https://github.com/kubernetes/enhancements/issues/4191

**Comment by kannon92** [comment]
/lgtm

cc @kubernetes/milestone-maintainers 


We should get this into 1.31.

**Comment by dims** [comment]
@kannon92 some of these are relevant to 1.29/1.30 as well? i see https://github.com/kubernetes/kubernetes/pull/120616 may have introduced some of these failure paths?

**Comment by kannon92** [comment]
> @kannon92 some of these are relevant to 1.29/1.30 as well? i see #120616 may have introduced some of these failure paths?

Yes, but from what I understand, we don't cherry-pick PRs for alpha features.

@saschagrunert how did you discover this bug? in 1.31, we are rolling this feature out as true.

**Comment by saschagrunert** [comment]
> > @kannon92 some of these are relevant to 1.29/1.30 as well? i see #120616 may have introduced some of these failure paths?
> 
> Yes, but from what I understand, we don't cherry-pick PRs for alpha features.
> 
> @saschagrunert how did you discover this bug? in 1.31, we are rolling this feature out as true.

@uablrek pointed out the issue when testing with CRI-O v1.29 and Kubernetes master. I can check wrt cherry picks when this one merges.

**Review by mrunalp — APPROVED** [review]

**Comment by saschagrunert** [comment]
Cherry-picks for 1.30 and 1.29 would apply:

- https://github.com/kubernetes/kubernetes/pull/126456
- https://github.com/kubernetes/kubernetes/pull/126457

**Comment by liggitt** [comment]
this was marked as a regression... can we pinpoint the specific minor version where this regressed and include it in the release note?

**Comment by saschagrunert** [comment]
> this was marked as a regression... can we pinpoint the specific minor version where this regressed and include it in the release note?

The code has been added in [26923b91](https://github.com/kubernetes/kubernetes/commit/26923b91e8cbd2d409e5d177ddd509429b76cb35#diff-3d39446eec61a856fae92b7001662519c7dc67ada5dc324546146bf4c4115709R242), which got released in v1.29. The code is only available when `features.KubeletSeparateDiskGC` is enabled, which was new in v1.28. Does this still work as regression?

**Pull request merged**
